### ISSUE TITLE
Maintained Documentation

### ISSUE THREAD

**Issue by felipeboffnunes** [google/trax#506]
### Description
Trax is a library for deep learning that focuses on sequence models and reinforcement learning. It combines performance with code clarity and maintained documentation and tests.
...

Sorry to bother, I'll be brief. I don't think the "maintained documentation" part of the statement is true (yet?). I like the work and I respect every project that goes deep down on neural network implementation, but I feel there is a critical lack of documentation for this project.

I was giving a look at Flax read the docs and, although the projects have different motives, I believe there should be something alike for Trax.

Again, sorry to bother. Wish all the luck and success for the project.

**Comment by jalammar** [comment]
Maybe this is something outside contributors can help with. To me, this also brings to mind the [discussions](https://twitter.com/srush_nlp/status/1262055375039344640) recently on Twitter. While some of these discussions are on API design and library scope, It has been making me think of ways the community can support with the docs for Trax.

One of my favorite ways to think of this is how "[What nobody tells you about documentation](https://documentation.divio.com/)" breaks it down into:
- Tutorials
- How-To Guides
- Explanation
- References

I suspect that at this point in time, references are the key required artifact. @lukaszkaiser [indicated](https://twitter.com/lukaszkaiser/status/1261471585124274177) the team is currently working on it. If the team is inclined, it could be worth their time to tag a few issues "good first issue" or "contributions welcome" for newcomers to start helping out.

**Current Trax Docs Inventory**
Besides README.md and the code API, the other documentation artifacts are currently:
[Trax Quick Intro Colab](https://colab.research.google.com/github/google/trax/blob/master/trax/intro.ipynb)
Short, sweet, and to the point. I really like how to-the-point this colab was. 

Reformer Colabs:
These are How-to guides for three tasks currently:
- [Image Generation](https://github.com/google/trax/blob/master/trax/models/reformer/image_generation.ipynb)
- [Machine Translation](https://github.com/google/trax/blob/master/trax/models/reformer/machine_translation.ipynb)
- [Text Generation](https://github.com/google/trax/blob/master/trax/models/reformer/text_generation.ipynb)

**What Docs Should be Produced Next?**
Curious to hear from the contributors and community what they feel is needed next.

Personally, it's important to know which aim of the library takes precedence:
A - The examples/models -- e.g. people coming in to use the Reformer, or continue where Transformer2Transformer left off, or for a transformer implementation that works on CPUs/GPUs/TPUs. Or;
B - to learn a new deep learning library from the ground up.

**Comment by cgarciae** [comment]
Same question. I was kind of expecting an API Reference given the statement.

**Comment by agrawalparth08** [comment]
Just the other day I was trying to visualize the model layers in a better way (like keras model summary ) and I found some moderate code help in intro.ipynb in layers (https://github.com/google/trax/blob/master/trax/layers/intro.ipynb) which is incomplete in the models part. Tried starting with this small cleaner function, may need more recursive way to pretty print/create a graphical layout of the model.
`def show_layer_properties(layer_obj, layer_name):
    template = ('{}_{}: {}')
    print("Total layers - ",len(layer_obj.sublayers))
    for i in range(len(layer_obj.sublayers)):
        print("---------")
        print(template.format(layer_name,i, layer_obj.sublayers[i]))
    
show_layer_properties(model,'Serial.sublayers')`
![image](https://user-images.githubusercontent.com/5490997/87386018-2ea0c380-c5bd-11ea-81fb-3d654bbef1bd.png)

**Comment by agrawalparth08** [comment]
@jalammar Just the other day I was trying to visualize the model layers in a better way (like keras model summary ) and I found some moderate code help in intro.ipynb in layers which is incomplete in the models part. Tried starting with this small cleaner function, may need more recursive way to pretty print/create a graphical tree like layout of the model which would help in debugging models as well.

![image](https://user-images.githubusercontent.com/5490997/87386018-2ea0c380-c5bd-11ea-81fb-3d654bbef1bd.png)

**Comment by jalammar** [comment]
That's wonderful, @lukaszkaiser! Congrats for shipping!

Since reading your comment, I have already started working on a Transformer colab for trax/model/transformer. I'll be sharing a draft in the next couple of days. I think I have a gentle narrative nailed down for it. If there's any additional guidance on what you and the team want to see in the colab, please let me know. I'm considering for the flow to start with a trained TransformerLM, build some confidence in the reader by establishing the model, the blocks, and the Trainer, then move to the Transformer model. I've always found it beneficial to readers if they go over a trained model first, then start learning about the training process once they're comfortable with the model.

For the other two (API docs for Conv/pooling and image models), maybe create an issue for each of the two (for visibility) and tag with "documentation", "good first issue" and "contributions welcome" (this repo has a "help wanted" tag, Tensorflow and JAX have "contributions welcome", either can work)?

**Comment by lukaszkaiser** [comment]
I created #897 for starters, please take it if you wish :). We also updated the main page README and the trax intro colab -- it has a direct Transformer decoding part with a pre-trained En-De model. So there is one trained model, I'm happy to add a LM too if you think having a decoder-only model would be helpful.

Will create more issues (and tags/labels) soon, thanks for the suggestions!!
